_posts/2017-08-18-Refactoring-with-reactJS-shared-state.md

```diff
@@ -0,0 +1,252 @@
+
+# Create the Redux Store
+
+First, we need to create a Redux store for our application. We're going to keep this piece with the React initilization script from part 1. Since the store needs to be accessible from all of our legacy code, we decided to create the store as a global. Please keep your 'boos' to a light rumble.
+
+The new code here is line `11`, where we create the store using our reducer (which we'll cover next) and lines `29`-`34` where we wrap each component we render in `react-redux`'s `Provider` component to give it access to the store.
+
+{% highlight javascript linenos %}
+import React from 'react';
+import ReactDOM from 'react-dom';
+import SuccessBanner from './successBanner.jsx';
+import FailureBanner from './failureBanner.jsx';
+import $ from 'jquery';
+import { createStore } from 'redux';
+import { Provider } from 'react-redux';
+import reducer from './reducers/index.jsx'
+
+// create the Redux store on the window to make accessible to whole application
+window.store = createStore(reducer);
+
+const componentRegistry = {
+    SuccessBanner,
+    FailureBanner
+};
+
+$(()=>{
+    const reactComponents = $('[data-react-component]');
+
+    $.each(reactComponents, function(index, obj) {
+        const props = $(obj).data();
+        const componentName = $(obj).data('reactComponent');
+        const Component = componentRegistry[componentName];
+
+        if (Component) {
+            // Render the component into the DOM
+            // wrapped in Provider to give it access to the Redux store
+            ReactDOM.render(
+                <Provider store={window.store}>
+                    <Component {...props} />
+                </Provider>,
+                obj
+            );
+        }
+    });
+});
+{% endhighlight %}
+
+Since we're interested in keeping track of the `FailureBanner`'s props in the application state, we wrote a simple reducer (that the store is created with). In this example, we assume that we only need to track the props for one `FailureBanner`. We store the `isShown` and `message` props as the application state. Any action dispatched with the type `SET_FAILURE_BANNER` updates the properties on the state.
+
+{% highlight javascript linenos %}
+const initialState = {
+    isShown: false,
+    message: ''
+};
+export default function(state = initialState, action) {
+    switch (action.type) {
+        case 'SET_FAILURE_BANNER':
+            return {
+                isShown: action.payload.isShown,
+                message: action.payload.message
+            };
+        default:
+            return state;
+    }
+}
+{% endhighlight %}
+
+Finally, we'll write an [action creator](https://daveceddia.com/redux-action-creators/) that creates the action object we want to dispatch when we want to change the Failure Banner state.
```

Comment:
This pattern allows for consistent readable code and ensures fewer typo based failures among other things.